File: dotnet/CucumberExpressions/Parsing/CucumberExpressionTokenizer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using CucumberExpressions.Ast;

namespace CucumberExpressions.Parsing;

public class CucumberExpressionTokenizer
{
    public Token[] Tokenize(string expression)
    {
        return TokenizeInternal(expression).ToArray();
    }
    public IEnumerable<Token> TokenizeInternal(string expression)
    {
        var iterator = new TokenIterator(expression);
        while (iterator.HasNext())
        {
            yield return iterator.Next();
        }
    }

    private class TokenIterator
    {
        private readonly string _expression;
        private readonly IEnumerator<int> _codePoints;

        private StringBuilder _buffer = new();
        private int _bufferCodePointCount;
        private TokenType _previousTokenType = TokenType.UNKNOWN;
        private TokenType _currentTokenType = TokenType.START_OF_LINE;
        private bool _treatAsText;
        private int _bufferStartIndex;
        private int _escaped;

        public TokenIterator(string expression)
        {
            _expression = expression;
            _codePoints = CodePointsOf(expression).GetEnumerator();
        }

        private Token ConvertBufferToToken(TokenType tokenType)
        {
            int escapeTokens = 0;
            if (tokenType == TokenType.TEXT)
            {
                escapeTokens = _escaped;
                _escaped = 0;
            }
            int consumedIndex = _bufferStartIndex + _bufferCodePointCount + escapeTokens;
            Token t = new Token(_buffer.ToString(), tokenType, _bufferStartIndex, consumedIndex);
            _buffer = new StringBuilder();
            _bufferCodePointCount = 0;
            this._bufferStartIndex = consumedIndex;
            return t;
        }

        private void AppendCodePoint(int codePoint)
        {
            _buffer.Append(char.ConvertFromUtf32(codePoint));
            _bufferCodePointCount++;
        }

        private void AdvanceTokenTypes()
        {
            _previousTokenType = _currentTokenType;
            _currentTokenType = TokenType.UNKNOWN;
        }

        private TokenType TokenTypeOf(int token, bool treatAsText)
        {
            if (!treatAsText)
            {
                return Token.TypeOf(token);
            }
            if (Token.CanEscape(token))
            {
                return TokenType.TEXT;
            }
            throw CucumberExpressionException.CreateCantEscape(_expression, _bufferStartIndex + _bufferCodePointCount + _escaped);
        }

        // Offsets are codepoint offsets, so the expression has to be walked a
        // codepoint at a time. A plain foreach over a string yields UTF-16 code
        // units, which counts anything outside the BMP twice. netstandard2.0
        // has no EnumerateRunes, hence the manual surrogate pairing.
        private static IEnumerable<int> CodePointsOf(string expression)
        {
            for (int i = 0; i < expression.Length; i++)
            {
                if (char.IsHighSurrogate(expression[i])
                    && i + 1 < expression.Length
                    && char.IsLowSurrogate(expression[i + 1]))
                {
                    yield return char.ConvertToUtf32(expression[i], expression[i + 1]);
                    i++;
                }
                else
                {
                    yield return expression[i];
                }
            }
        }

        private bool ShouldContinueTokenType(TokenType? previousTokenType, TokenType? currentTokenType)
        {
            return currentTokenType == previousTokenType
                    && (currentTokenType == TokenType.WHITE_SPACE || currentTokenType == TokenType.TEXT);
        }

        public bool HasNext()
        {
            return _previousTokenType != TokenType.END_OF_LINE;
        }

        public Token Next()
        {
            if (!HasNext())
            {
                throw new InvalidOperationException("no such element");
            }
            if (_currentTokenType == TokenType.START_OF_LINE)
            {
                Token token = ConvertBufferToToken(_currentTokenType);
                AdvanceTokenTypes();
                return token;
            }

            while (_codePoints.MoveNext())
            {
                int codePoint = _codePoints.Current;
                if (!_treatAsText && Token.IsEscapeCharacter(codePoint))
                {
                    _escaped++;
                    _treatAsText = true;
                    continue;
                }
                _currentTokenType = TokenTypeOf(codePoint, _treatAsText);
                _treatAsText = false;

                if (_previousTokenType == TokenType.START_OF_LINE ||
                        ShouldContinueTokenType(_previousTokenType, _currentTokenType))
                {
                    AdvanceTokenTypes();
                    AppendCodePoint(codePoint);
                }
                else
                {
                    Token t = ConvertBufferToToken(_previousTokenType);
                    AdvanceTokenTypes();
                    AppendCodePoint(codePoint);
                    return t;
                }
            }

            if (_bufferCodePointCount > 0)
            {
                Token previousToken = ConvertBufferToToken(_previousTokenType);
                AdvanceTokenTypes();
                return previousToken;
            }

            _currentTokenType = TokenType.END_OF_LINE;
            if (_treatAsText)
            {
                throw CucumberExpressionException.CreateTheEndOfLineCanNotBeEscaped(_expression);
            }
            Token currentToken = ConvertBufferToToken(_currentTokenType);
            AdvanceTokenTypes();
            return currentToken;
        }

    }

}

File: elixir/lib/cucumber/cucumber_expressions/cucumber_expression_tokenizer.ex
defmodule Cucumber.CucumberExpressions.CucumberExpressionTokenizer do
  @moduledoc """
  Splits a Cucumber Expression source string into a list of
  `Cucumber.CucumberExpressions.Token`s.
  """

  alias Cucumber.CucumberExpressions.Error
  alias Cucumber.CucumberExpressions.Token

  @spec tokenize(String.t()) :: {:ok, [Token.t()]} | {:error, Error.t()}
  def tokenize(expression) do
    codepoints = String.to_charlist(expression)

    state = %{
      tokens: initial_tokens(codepoints),
      buffer: [],
      buffer_start_index: 0,
      previous_token_type: :start_of_line,
      treat_as_text: false,
      escaped: 0
    }

    loop(codepoints, expression, length(codepoints), state)
  end

  @spec tokenize!(String.t()) :: [Token.t()]
  def tokenize!(expression) do
    case tokenize(expression) do
      {:ok, tokens} -> tokens
      {:error, error} -> raise error
    end
  end

  defp initial_tokens([]), do: [%Token{type: :start_of_line, text: "", start: 0, end: 0}]
  defp initial_tokens(_), do: []

  defp loop([codepoint | rest], expression, total, %{treat_as_text: false} = state) do
    if Token.escape_character?(codepoint) do
      loop(rest, expression, total, %{state | escaped: state.escaped + 1, treat_as_text: true})
    else
      consume(codepoint, rest, expression, total, state)
    end
  end

  defp loop([codepoint | rest], expression, total, state) do
    consume(codepoint, rest, expression, total, state)
  end

  defp loop([], expression, total, state) do
    state =
      if state.buffer == [] do
        state
      else
        {token, state} = convert_buffer_to_token(state, state.previous_token_type)
        %{state | tokens: [token | state.tokens]}
      end

    if state.treat_as_text do
      {:error, Error.end_of_line_cannot_be_escaped(expression)}
    else
      end_of_line = %Token{type: :end_of_line, text: "", start: total, end: total}
      {:ok, Enum.reverse([end_of_line | state.tokens])}
    end
  end

  defp consume(codepoint, rest, expression, total, state) do
    case token_type_of(codepoint, state, expression) do
      {:error, _} = error ->
        error

      {:ok, current_token_type} ->
        state = %{state | treat_as_text: false}

        state =
          if should_create_new_token?(current_token_type, state.previous_token_type) do
            {token, state} = convert_buffer_to_token(state, state.previous_token_type)

            %{
              state
              | tokens: [token | state.tokens],
                previous_token_type: current_token_type,
                buffer: [codepoint]
            }
          else
            %{state | previous_token_type: current_token_type, buffer: [codepoint | state.buffer]}
          end

        loop(rest, expression, total, state)
    end
  end

  defp token_type_of(codepoint, %{treat_as_text: false}, _expression) do
    {:ok, Token.type_of(codepoint)}
  end

  defp token_type_of(codepoint, %{treat_as_text: true} = state, expression) do
    if Token.can_escape?(codepoint) do
      {:ok, :text}
    else
      index = state.buffer_start_index + length(state.buffer) + state.escaped
      {:error, Error.cant_escape(expression, index)}
    end
  end

  defp should_create_new_token?(current_token_type, previous_token_type) do
    current_token_type != previous_token_type or
      (current_token_type != :white_space and current_token_type != :text)
  end

  defp convert_buffer_to_token(
         %{buffer: buffer, buffer_start_index: start, escaped: escaped} = state,
         type
       ) do
    escape_tokens = if type == :text, do: escaped, else: 0
    consumed_index = start + length(buffer) + escape_tokens

    token = %Token{
      type: type,
      text: List.to_string(Enum.reverse(buffer)),
      start: start,
      end: consumed_index
    }

    escaped = if type == :text, do: 0, else: escaped
    {token, %{state | buffer: [], buffer_start_index: consumed_index, escaped: escaped}}
  end
end
